protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+		Ok(EthClient {
+			rpc_provider: args.rpc_provider,
+			chain_id: args.chain_id,
+			ws_provider: args.ws_provider,
+			signing_address: args.signing_address,
+			initiator_address: args.initiator_address,
+			gas_limit: args.gas_limit,
+			num_tx_send_retries: args.num_tx_send_retries,
+			send_transaction_error_rules,
+		})
+	}
+}
+
+impl<P> Clone for EthClient<P> {
+	fn clone(&self) -> Self {
+		todo!()
+	}
+}
+
+#[async_trait::async_trait]
+impl<P> BridgeContractInitiator for EthClient<P>
+where
+	P: Provider + Clone + Send + Sync + Unpin,
+{
+	type Address = EthAddress;
+	type Hash = EthHash;
+
+	async fn initiate_bridge_transfer(
+		&mut self,
+		initiator_address: InitiatorAddress<Self::Address>,
+		recipient_address: RecipientAddress<Vec<u8>>,
+		hash_lock: HashLock<Self::Hash>,
+		time_lock: TimeLock,
+		amount: Amount,
+	) -> BridgeContractInitiatorResult<()> {
+		let contract = AtomicBridgeInitiator::new(initiator_address.0 .0, &self.rpc_provider);
+		let recipient_bytes: [u8; 32] = recipient_address.0.try_into().unwrap();
+		let call = contract.initiateBridgeTransfer(
+			U256::from(amount.0),
+			FixedBytes(recipient_bytes),
+			FixedBytes(hash_lock.0),
+			U256::from(time_lock.0),
+		);
+		let _ = send_transaction(
+			call,
+			&self.send_transaction_error_rules,
+			self.num_tx_send_retries,
+			self.gas_limit as u128,
+		)
+		.await;
+		Ok(())
+	}
+
+	async fn complete_bridge_transfer(
+		&mut self,
+		bridge_transfer_id: BridgeTransferId<Self::Hash>,
+		pre_image: HashLockPreImage,
+	) -> BridgeContractInitiatorResult<()> {
+		let pre_image: [u8; 32] = utils::vec_to_array(pre_image.0)?;
+		let contract = AtomicBridgeInitiator::new(self.signing_address.0, &self.rpc_provider);
+		let call = contract
+			.completeBridgeTransfer(FixedBytes(bridge_transfer_id.0), FixedBytes(pre_image));
+		let _ = send_transaction(
+			call,
+			&self.send_transaction_error_rules,
+			self.num_tx_send_retries,
+			self.gas_limit as u128,
+		)
+		.await;
+		Ok(())
+	}
+
+	async fn refund_bridge_transfer(
+		&mut self,
+		bridge_transfer_id: BridgeTransferId<Self::Hash>,
+	) -> BridgeContractInitiatorResult<()> {
+		let contract = AtomicBridgeInitiator::new(self.signing_address.0, &self.rpc_provider);
+		let call = contract.refundBridgeTransfer(FixedBytes(bridge_transfer_id.0));
+		let _ = send_transaction(
+			call,
+			&self.send_transaction_error_rules,
+			self.num_tx_send_retries,
+			self.gas_limit as u128,
+		)
+		.await;
+		Ok(())
+	}
+
+	async fn get_bridge_transfer_details(
+		&mut self,
+		bridge_transfer_id: BridgeTransferId<Self::Hash>,
+	) -> BridgeContractInitiatorResult<Option<BridgeTransferDetails<Self::Address, Self::Hash>>> {
+		let mapping_slot = U256::from(0); // the mapping is the zeroth slot in the contract
+		let key = bridge_transfer_id.0;
+		let storage_slot = self.calculate_storage_slot(key, mapping_slot);
+		let storage: U256 = self
+			.rpc_provider
+			.get_storage_at(self.initiator_address.0, storage_slot)
```

Comment:
This makes me think this is the intiator contract address not the inititiator address which makes sense.